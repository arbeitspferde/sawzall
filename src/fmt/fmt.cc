// Copyright 2010 Google Inc.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ------------------------------------------------------------------------

/*
 * The authors of this software are Rob Pike and Ken Thompson.
 *              Copyright (c) 2002 by Lucent Technologies.
 * Permission to use, copy, modify, and distribute this software for any
 * purpose without fee is hereby granted, provided that this entire notice
 * is included in all copies of any software which is or includes a copy
 * or modification of this software and in all copies of the supporting
 * documentation for such software.
 * THIS SOFTWARE IS BEING PROVIDED "AS IS", WITHOUT ANY EXPRESS OR IMPLIED
 * WARRANTY.  IN PARTICULAR, NEITHER THE AUTHORS NOR LUCENT TECHNOLOGIES MAKE ANY
 * REPRESENTATION OR WARRANTY OF ANY KIND CONCERNING THE MERCHANTABILITY
 * OF THIS SOFTWARE OR ITS FITNESS FOR ANY PARTICULAR PURPOSE.
 */
#include <stdarg.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include "fmt/fmt.h"
#include "fmt/fmtdef.h"

namespace Fmt {

Formatter _stdfmt;

Formatter::Formatter() {
	this->nfmt = 0;
	memset(this->fmt, 0, sizeof this->fmt);
}

Formatter::~Formatter() {
}

static Convfmt knownfmt[] = {
	{ ' ',	 true, false, __flagfmt		},
	{ '#',	 true, false, __flagfmt		},
	{ '%',	false, false, __percentfmt	},
	{ '\'',	 true, false, __flagfmt		},
	{ '+',	 true, false, __flagfmt		},
	{ '-',	 true, false, __flagfmt		},
	{ 'C',	false,  true, __runefmt		},	/* Plan 9 addition */
	{ 'E',	false,  true, __efgfmt		},
	{ 'F',	false,  true, __efgfmt		},	/* ANSI only */
	{ 'G',	false,  true, __efgfmt		},
	{ 'L',	 true, false, __flagfmt		},	/* ANSI only */
	{ 'S',	false,  true, __runesfmt	},	/* Plan 9 addition */
	{ 'X',	false,  true, __ifmt		},
	{ 'b',	false,  true, __ifmt		},	/* Plan 9 addition */
	{ 'c',	false,  true, __charfmt		},
	{ 'd',	false,  true, __ifmt		},
	{ 'e',	false,  true, __efgfmt		},
	{ 'f',	false,  true, __efgfmt		},
	{ 'g',	false,  true, __efgfmt		},
	{ 'h',	 true, false, __flagfmt		},
	{ 'i',	false,  true, __ifmt		},	/* ANSI only */
	{ 'l',	 true, false, __flagfmt		},
	{ 'n',	false,  true, __countfmt	},
	{ 'o',	false,  true, __ifmt		},
	{ 'p',	false,  true, __ifmt		},
	{ 'r',	false, false, __errfmt		},
	{ 's',	false,  true, __strfmt		},
	{ 'u',	false,  true, __ifmt		},	/* in Plan 9, __flagfmt */
	{ 'x',	false,  true, __ifmt		},
	{ 0,	false, false, nil		},
};

static Convfmt badconvfmt = { 0, false, false, __badfmt };

static int
doinstallfmt(Formatter* formatter, int c, bool isflag, bool consumesarg, Fmts f)
{
	Convfmt *p, *ep;

	if(c<=0 || c>=65536)
		return -1;
	if(!f)
		f = __badfmt;

	ep = &formatter->fmt[formatter->nfmt];
	for(p=formatter->fmt; p<ep; p++)
		if(p->c == c)
			break;

	if(p == &formatter->fmt[Maxfmt])
		return -1;

	if(p->fmt && p->fmt != f){
		/* note: this test only catches an explicit install; if someone
		 * overrides %d and then someone else expects %d to be the
		 * default, the second someone will be surprised.  to be fixed.
		 */
		/* print without calling install */
		int n;
		Rune r;
		char m1[] = "Fmt: Warning: verb %";
		char m2[UTFmax+1];
		r = c;
		n = runetochar(m2, &r);
		m2[n] = '\0';
		char m3[] =  " redefined\n";
		write(2, m1, strlen(m1));
		write(2, m2, strlen(m2));
		write(2, m3, strlen(m3));
	}
	p->fmt = f;
	p->isflag = isflag;
	p->consumesarg = consumesarg;
	if(p == ep){ /* installing a new format character */
		formatter->nfmt++;
		p->c = c;
	}

	return 0;
}

int
Formatter::installverb(int c, bool consumesarg, Fmts f)
{
	int ret;

	ret = doinstallfmt(this, c, false, consumesarg, f);
	return ret;
}

int
Formatter::installflag(int c, bool consumesarg, Fmts f)
{
	int ret;

	ret = doinstallfmt(this, c, true, consumesarg, f);
	return ret;
}

Convfmt*
fmtfmt(Formatter* formatter, int c)
{
	Convfmt *p, *ep;

	ep = &formatter->fmt[formatter->nfmt];
	for(p=formatter->fmt; p<ep; p++)
		if(p->c == c)
			return p;

	/* is this a predefined format char? */
	for(p=knownfmt; p->c; p++)
		if(p->c == c)
			return p;

	return &badconvfmt;
}

static const void*
nextrune(const void *fmt, int isrunes, Rune *runep) {
	if(isrunes){
		*runep = *(Rune*)fmt;
		fmt = (Rune*)fmt + 1;
	}else
		fmt = (char*)fmt + chartorune(runep, (char*)fmt);
	return fmt;
}

// *runep already holds next rune; always sets it to 0 if no number
static const void*
nextnumber(const void *fmt, int isrunes, Rune *runep, int *intp){
	int i;
	
	i = 0;
	while('0' <= *runep && *runep <= '9'){
		i = i * 10 + *runep - '0';
		fmt = nextrune(fmt, isrunes, runep);
	}
	*intp = i;
	return fmt;
}

/*
 * always sets *dollarp to 0 if no $ present.
 * *runep has not yet been read or consumed.
 * upon return, *runep contains the next rune.
 */
static const void*
getdollar(const void *fmt, int isrunes, Rune *runep, int *dollarp){
	int oldrune;

	fmt = nextrune(fmt, isrunes, runep);
	const void *oldfmt = fmt;
	oldrune = *runep;
	fmt = nextnumber(fmt, isrunes, runep, dollarp);
	if(*runep != '$'){
		*dollarp = 0;
		*runep = oldrune;
		return oldfmt;
	}
	return nextrune(fmt, isrunes, runep);
}

static const void**
addverb(const void **verbs, int *nverbs,  int argno, const void *fmt)
{
	const void** oldverbs;
	int n;

	if(verbs == nil || argno >= *nverbs){
		n = 2 * (argno + 1);
		oldverbs = verbs;
		verbs = (const void**)realloc(verbs, n * sizeof verbs[0]);
		if(verbs == nil){
			free(oldverbs);
			return nil;
		}
		memset(verbs + *nverbs, 0, (n - *nverbs) * sizeof verbs[0]);
		*nverbs = n;
	}
	/*
	 * Needed only to consume arguments.  If there are two verbs for
	 * a given argument, they need to consume the same number of
	 * words or the user deserves whatever happens.
	 */
	if(verbs[argno] == nil)
		verbs[argno] = fmt;
	return verbs;
}

/* dofmt for a format that includes %1$d -style stuff */
static int
doreorderfmt(Formatter* formatter, State *f, const void *fmt, int isrunes, va_list* args, int nargs)
{
	Rune *rt, *rs;
	int r;
	char *t, *s;

	for(;;){
		if(f->runes){
			rt = (Rune*)f->to;
			rs = (Rune*)f->stop;
			for(;;){
				fmt = nextrune(fmt, isrunes, &r);
				if(r == '\0'){
					f->nfmt += rt - (Rune *)f->to;
					f->to = rt;
					return 1;
				}
				if(r == '%'){
					break;
				}
				FMTRCHAR(f, rt, rs, r);
			}
			f->nfmt += rt - (Rune *)f->to;
			f->to = rt;
		}else{
			t = (char*)f->to;
			s = (char*)f->stop;
			for(;;){
				fmt = nextrune(fmt, isrunes, &r);
				if(r == '\0'){
					f->nfmt += t - (char *)f->to;
					f->to = t;
					return 1;
				}
				if(r == '%'){
					break;
				}
				if(r < Runeself){
					FMTCHAR(f, t, s, r);
				}else{
					FMTRUNE(f, t, s, r);
				}
			}
			f->nfmt += t - (char *)f->to;
			f->to = t;
		}

		fmt = dispatch(formatter, f, fmt, isrunes, args, nargs);
		if(fmt == nil){
			return -1;
		}
	}

	return 1;
}

/*
 * Call va_end on the first nargs elements of args and free args
 */
void cleanupargs(va_list* args, int nargs) {
	int a;

	for(a = 1; a < nargs; a++){
		va_end(args[a]);
	}
	free(args);
}

/*
 * 'Fmt' has some %1$d -style formats; it points right after the '%' for the
 * first one.  Find all arguments and their locations, then produce the
 * formatted output.  Returns the empty string or nil if an error is found.
 */
static Rune emptyrunes[1];
static const void*
reorderfmt(Formatter* formatter, State* ff, const void *fmt, int isrunes)
{
	State nullf;
	va_list* args;
	Rune r;
	const void *origfmt, *argstart, **verbs;
	int a, i, allocedv, argno, maxverb, maxflag, nargs, dollar, star;

	/*
	 * backup over the preceding %
	 */
	if(isrunes)
		fmt = (Rune*)fmt - 1;
	else
		fmt = (char*)fmt - 1;

	/*
	 * Build a vector of verbs in argument order.
	 *
	 * Ignore any width or precision flags.  Later we'll assume all missing
	 * args are such flags.
	 */
	origfmt = fmt;
	verbs = nil;
	allocedv = 0;
	maxverb = 0;
	for (;;) {
		fmt = nextrune(fmt, isrunes, &r);
		if(r == '\0')
			break;
		if(r != '%')
			continue;

		/*
		 * look for potential number$
		 */
		argstart = fmt;
		fmt = nextrune(fmt, isrunes, &r);
		fmt = nextnumber(fmt, isrunes, &r, &argno);
		if(r != '$' || argno == 0){
			/*
			 * not valid $ verb syntax, assume the verb
			 * doesn't consume an arg
			 */
			continue;
		}

		verbs = addverb(verbs, &allocedv, argno, argstart);
		if(verbs == nil){
			return nil;
		}
		if(argno >= maxverb)
			maxverb = argno + 1;
	}

	/*
	 * Allocate enough args for a width and precision flag for every verb
	 */
	nargs = 3 * maxverb;	/* verb + width + precision -> 3 args per verb */
	if(nargs < 2)
		nargs = 2;
	args = (va_list*)malloc(nargs * sizeof(va_list));
	if(args == nil){
		free(verbs);
		return nil;
	}

	/*
	 * Run over each of the formatting verbs in argument order
	 * and record their va_list positions.
	 */
	formatter->fmtnullinit(&nullf);
	nullf.args = ff->args;
	maxflag = maxverb;
	if(maxflag < 2)
		maxflag = 2;
	for(a = 1; a < maxverb; ) {
		VA_COPY(args[a], *nullf.args);
		fmt = verbs[a];
		a++;
		if(fmt == nil){		/* a width or precision flag */
			va_arg(*nullf.args, int);
			continue;
		}

		fmt = getdollar(fmt, isrunes, &r, &dollar);
		nullf.flags = 0;
		nullf.width = nullf.prec = 0;
		for(;; fmt = nextrune(fmt, isrunes, &r)){
		 HaveNextRune:
			// at this point, r is always the next character
			switch(r){
			case '\0':
				cleanupargs(args, a);
				free(verbs);
				return nil;
			case '.':
				nullf.flags |= Fmt::FmtWidth|Fmt::FmtPrec;
				continue;
			case '0':
				if(!(nullf.flags & Fmt::FmtWidth)){
					nullf.flags |= Fmt::FmtZero;
					fmt = nextrune(fmt, isrunes, &r);
					continue;
				}
				/* fall through */
			case '1': case '2': case '3': case '4':
			case '5': case '6': case '7': case '8': case '9':
				fmt = nextnumber(fmt, isrunes, &r, &i);
			numflag:
				if(nullf.flags & Fmt::FmtWidth){
					nullf.flags |= Fmt::FmtPrec;
					nullf.prec = i;
				}else{
					nullf.flags |= Fmt::FmtWidth;
					nullf.width = i;
				}
				goto HaveNextRune;
			case '*':
				fmt = getdollar(fmt, isrunes, &r, &star);
				if(star >= maxflag){
					maxflag = star + 1;
				}

				/* can safely make up any value. */
				i = 17;
				goto numflag;
			}
			nullf.r = r;
			Convfmt *convfmt = fmtfmt(formatter, r);
			if((*convfmt->fmt)(&nullf) < 0){
				cleanupargs(args, a);
				free(verbs);
				return nil;
			}
			if(!convfmt->isflag){
				break;
			}
			/* a flag; $ arg number is implicit, keep going. */
		}
	}
	free(verbs);

	/*
	 * Can't have more flags than allowed for above unless there is an
	 * error in the format string.
	 */
	if(maxflag > nargs){
		cleanupargs(args, maxverb);
		return nil;
	}

	/*
	 * Fill in the arg location of any flags at the end.
	 */
	for(; a < maxflag; a++){
		VA_COPY(args[a], *nullf.args);
		va_arg(*nullf.args, int);
	}

	ff->args = nil;
	r = doreorderfmt(formatter, ff, origfmt, isrunes, args, maxflag);
	ff->args = nullf.args;

	cleanupargs(args, maxflag);
	if(r < 0)
		return nil;
	if(isrunes)
		return emptyrunes;
	return "";
}

const void*
dispatch(Formatter* formatter, State *f, const void *fmt, int isrunes, va_list* args, int nargs)
{
        va_list arg, stararg;
	const void* origfmt;
	Rune r;
	int dollar, star;

	f->flags = 0;
	f->width = f->prec = 0;

	origfmt = fmt;
	fmt = getdollar(fmt, isrunes, &r, &dollar);

	/*
	 * If using %1$d -style formats, make a copy of the argument.
	 * Need to make a copy since the argument can be used multiple times
	 * in the format string.
	 */
	if(dollar){
		/* May be the first %1$d -style arg seen. */
		if(args == nil){
			return reorderfmt(formatter, f, origfmt, isrunes);
		}
		if(dollar >= nargs){
			return nil;
		}
		VA_COPY(arg, args[dollar]);
		f->args = &arg;
	}
	for(;; fmt = nextrune(fmt, isrunes, &r)){
                int i = 0;
	 HaveNextRune:
		/* at this point, r is always the next character */
		switch(r){
		case '\0':
			fmt = nil;
			goto breakout;
		case '.':
			f->flags |= Fmt::FmtWidth|Fmt::FmtPrec;
			continue;
		case '0':
			if(!(f->flags & Fmt::FmtWidth)){
				f->flags |= Fmt::FmtZero;
				continue;
			}
			/* fall through */
		case '1': case '2': case '3': case '4':
		case '5': case '6': case '7': case '8': case '9':
			fmt = nextnumber(fmt, isrunes, &r, &i);
		numflag:
			if(f->flags & Fmt::FmtWidth){
				f->flags |= Fmt::FmtPrec;
				f->prec = i;
			}else{
				f->flags |= Fmt::FmtWidth;
				f->width = i;
			}
			goto HaveNextRune;
		case '*':
			fmt = getdollar(fmt, isrunes, &r, &star);
			if(args == nil){
				if(star != 0){
					return reorderfmt(formatter, f, origfmt, isrunes);
				}
				i = FMT_ARG(f, int);
			}else if(star <= 0){
				fmt = nil;		/* illegal */
				goto breakout;
			}else{
				/*
				 * reorderfmt can miss numbered '*' arguments
				 * for verbs that don't consume args, so we may
				 * have to advance past the end of the known
				 * argument positions.
				 */
				int a = star < nargs ? star : nargs - 1;
				VA_COPY(stararg, args[a]);
				for(; a <= star; a++)
					i = va_arg(stararg, int);
				va_end(stararg);
			}
			if(i < 0){
				/*
				 * negative precision =>
				 * ignore the precision.
				 */
				if(f->flags & Fmt::FmtPrec){
					f->flags &= ~Fmt::FmtPrec;
					f->prec = 0;
					goto HaveNextRune;
				}
				i = -i;
				f->flags |= Fmt::FmtLeft;
			}
			goto numflag;
		}

		f->r = r;
		Convfmt *convfmt = fmtfmt(formatter, r);

		/*
		 * If the formatter consumes an argument there must be one
		 * specified when using %1$d -style formats.
		 */
		if(args != nil && !dollar && convfmt->consumesarg){
			fmt = nil;
			break;
		}
		if((*convfmt->fmt)(f) < 0){
			/* Formatter returned an error */
			fmt = nil;
			break;
		}
		if(!convfmt->isflag){
			break;
		}
		/* else it's a flag; keep going */
	}
 breakout:
	if(args != nil && dollar){
		va_end(arg);
		f->args = nil;
	}
	return fmt;
}

}  // end namespace Fmt
