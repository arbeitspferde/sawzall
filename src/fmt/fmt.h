// Copyright 2010 Google Inc.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ------------------------------------------------------------------------


/*
 * The authors of this software are Rob Pike and Ken Thompson.
 *              Copyright (c) 2002 by Lucent Technologies.
 * Permission to use, copy, modify, and distribute this software for any
 * purpose without fee is hereby granted, provided that this entire notice
 * is included in all copies of any software which is or includes a copy
 * or modification of this software and in all copies of the supporting
 * documentation for such software.
 * THIS SOFTWARE IS BEING PROVIDED "AS IS", WITHOUT ANY EXPRESS OR IMPLIED
 * WARRANTY.  IN PARTICULAR, NEITHER THE AUTHORS NOR LUCENT TECHNOLOGIES MAKE ANY
 * REPRESENTATION OR WARRANTY OF ANY KIND CONCERNING THE MERCHANTABILITY
 * OF THIS SOFTWARE OR ITS FITNESS FOR ANY PARTICULAR PURPOSE.
 */

#ifndef _FMTH_
#define _FMTH_ 1

#include <stdarg.h>

#ifndef _UTFH_
#include "fmt/runes.h"
#endif

/*
 * The Fmt class is mostly a name space to capture the C-like interface
 * to the usual print, sprint, etc. routines.  Most programs can use it
 * without worrying about the Formatter class.
 *
 * The Formatter class allows the installation of custom print verbs.
 * The namesake routines in Fmt are implemented by calls to the
 * methods in Formatter, using a statically defined, read-only Formatter.
 */


/* This namespace provides lexical compatibility with the old class Fmt */
namespace Fmt {

class Formatter;

struct State{
	unsigned char	runes;	/* output buffer is runes or chars? */
	void	*start;			/* of buffer */
	void	*to;			/* current place in the buffer */
	void	*stop;			/* end of the buffer; overwritten if flush fails */
	int	(*flush)(State *);	/* called when to == stop */
	void	*farg;			/* to make flush a closure */
	int	fintarg;		/* holds ints used by flush routines */
	int	nfmt;			/* num chars formatted so far */
	va_list	*args;			/* args passed to verb */
	int	r;			/* % format Rune */
	int	width;
	int	prec;
	unsigned long	flags;
	Formatter	*formatter;
	const char *decimal;		/* representation of decimal point; cannot be "" */
	/* for %'d */
	const char *thousands;		/* separator for thousands */

	/* Each char is an integer indicating #digits before next separator. Values:
	 *	\xFF: no more grouping (or \x7F; defined to be CHAR_MAX in POSIX)
	 *	\x00: repeat previous indefinitely
	 *	\x**: count that many
	 */
	const char *grouping;		/* descriptor of separator placement */
};

enum{
	FmtWidth	= 1,
	FmtLeft		= FmtWidth << 1,
	FmtPrec		= FmtLeft << 1,
	FmtSharp	= FmtPrec << 1,
	FmtSpace	= FmtSharp << 1,
	FmtSign		= FmtSpace << 1,
	FmtApost	= FmtSign << 1,
	FmtZero		= FmtApost << 1,
	FmtUnsigned	= FmtZero << 1,
	FmtShort	= FmtUnsigned << 1,
	FmtLong		= FmtShort << 1,
	FmtVLong	= FmtLong << 1,
	FmtByte		= FmtVLong << 1,
	FmtLDouble	= FmtByte << 1,

	FmtFlag		= FmtLDouble << 1
};

enum
{
	Maxfmt = 64
};

typedef int (*Fmts)(State*);

struct Convfmt
{
	int		c;
	bool		isflag;		/* is the installed fmt a flag or a verb? */
	bool		consumesarg;	/* does it consume a vararg? */
	volatile Fmts	fmt;		/* for spin lock in fmtfmt; avoids race due to write order */
};

class Formatter {
public:
	Formatter();
	~Formatter();

	int	print(const char*, ...);
	char*	seprint(char*, char*, const char*, ...);
	char*	vseprint(char*, char*, const char*, va_list*);
	int	snprint(char*, int, const char*, ...);
	int	vsnprint(char*, int, const char*, va_list*);
	char*	smprint(const char*, ...);
	char*	vsmprint(const char*, va_list*);
	int	sprint(char*, const char*, ...);
	int	fprint(int, const char*, ...);
	int	vfprint(int, const char*, va_list*);

	int	runesprint(Rune*, const char*, ...);
	int	runesnprint(Rune*, int, const char*, ...);
	int	runevsnprint(Rune*, int, const char*, va_list*);
	Rune*	runeseprint(Rune*, Rune*, const char*, ...);
	Rune*	runevseprint(Rune*, Rune*, const char*, va_list*);
	Rune*	runesmprint(const char*, ...);
	Rune*	runevsmprint(const char*, va_list*);

	static void	fmtlocaleinit(State*, const char*, const char*, const char*);

	int	fmtnullinit(State*);
	int	fmtfdinit(State*, int, char*, int);
	int	fmtstrinit(State*);
	int	runefmtstrinit(State*);

	static int	fmtfdflush(State*);
	static char*	fmtstrflush(State*);
	static Rune*	runefmtstrflush(State*);

	void	quoteinstall(void);

	int	installverb(int c, bool consumesarg, int (*f)(State*));
	int	installflag(int c, bool consumesarg, int (*f)(State*));

	int	fmtprint(State*, const char*, ...);
	int	fmtvprint(State*, const char*, va_list*);

	int	nfmt;
	Convfmt	fmt[Maxfmt];
};

extern Formatter _stdfmt;

int	print(const char*, ...);
char*	seprint(char*, char*, const char*, ...);
char*	vseprint(char*, char*, const char*, va_list*);
int	snprint(char*, int, const char*, ...);
int	vsnprint(char*, int, const char*, va_list*);
char*	smprint(const char*, ...);
char*	vsmprint(const char*, va_list*);
int	sprint(char*, const char*, ...);
int	fprint(int, const char*, ...);
int	vfprint(int, const char*, va_list*);

int	runesprint(Rune*, const char*, ...);
int	runesnprint(Rune*, int, const char*, ...);
int	runevsnprint(Rune*, int, const char*, va_list*);
Rune*	runeseprint(Rune*, Rune*, const char*, ...);
Rune*	runevseprint(Rune*, Rune*, const char*, va_list*);
Rune*	runesmprint(const char*, ...);
Rune*	runevsmprint(const char*, va_list*);

int	fmtnullinit(State*);
int	fmtfdinit(State*, int, char*, int);
int	fmtfdflush(State*);
int	fmtstrinit(State*);
void	fmtlocaleinit(State*, const char*, const char*, const char*);
char*	fmtstrflush(State*);
Rune*	runefmtstrflush(State*);
int	runefmtstrinit(State*);

int	quotestrfmt(State *f);
int	quoterunestrfmt(State *f);
void	quoteinstall(void);
extern int	(*fmtdoquote)(int);


/*
 * dofmt -- format to a buffer
 * the number of characters formatted is returned,
 * or -1 if there was an error.
 * if the buffer is ever filled, flush is called.
 * it should reset the buffer and return whether formatting should continue.
 */
int	dofmt(Formatter*, State*, const char*);
int	dorfmt(Formatter*, State*, const Rune*);
int	fmtprint(State*, const char*, ...);
int	fmtvprint(State*, const char*, va_list*);
int	fmtrune(State*, int);
int	fmtstrcpy(State*, const char*);
int	fmtrunestrcpy(State *f, const Rune *s);

double	fmtstrtod(const char *, char **);
double	fmtcharstod(int(*)(void*), void*);

};

#define	FMT_ARG(f, type) va_arg(*(f)->args, type)

#endif
