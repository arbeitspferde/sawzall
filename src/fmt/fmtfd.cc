// Copyright 2010 Google Inc.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ------------------------------------------------------------------------

/*
 * The authors of this software are Rob Pike and Ken Thompson.
 *              Copyright (c) 2002 by Lucent Technologies.
 * Permission to use, copy, modify, and distribute this software for any
 * purpose without fee is hereby granted, provided that this entire notice
 * is included in all copies of any software which is or includes a copy
 * or modification of this software and in all copies of the supporting
 * documentation for such software.
 * THIS SOFTWARE IS BEING PROVIDED "AS IS", WITHOUT ANY EXPRESS OR IMPLIED
 * WARRANTY.  IN PARTICULAR, NEITHER THE AUTHORS NOR LUCENT TECHNOLOGIES MAKE ANY
 * REPRESENTATION OR WARRANTY OF ANY KIND CONCERNING THE MERCHANTABILITY
 * OF THIS SOFTWARE OR ITS FITNESS FOR ANY PARTICULAR PURPOSE.
 */
#include <stdarg.h>
#include <string.h>
#include "fmt/fmt.h"
#include "fmt/fmtdef.h"

namespace Fmt {

/*
 * public routine for final flush of a formatting buffer
 * to a file descriptor; returns total char count.
 */
int
Formatter::fmtfdflush(State *f)
{
	if(__fmtFdFlush(f) <= 0)
		return -1;
	return f->nfmt;
}

int
fmtfdflush(State *f)
{
	return Formatter::fmtfdflush(f);
}

/*
 * initialize an output buffer for buffered printing
 */
int
Formatter::fmtfdinit(State *f, int fd, char *buf, int size)
{
	f->runes = 0;
	f->start = buf;
	f->to = buf;
	f->stop = buf + size;
	f->flush = __fmtFdFlush;
	f->farg = 0;
	f->fintarg = fd;
	f->flags = 0;
	f->nfmt = 0;
	f->formatter = this;
	fmtlocaleinit(f, nil, nil, nil);
	return 0;
}

int
fmtfdinit(State *f, int fd, char *buf, int size)
{
	return _stdfmt.fmtfdinit(f, fd, buf, size);
}

}  /* end namespace Fmt */
