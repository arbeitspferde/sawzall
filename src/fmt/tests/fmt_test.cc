// Copyright 2010 Google Inc.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ------------------------------------------------------------------------

/*
 * The authors of this software are Rob Pike and Ken Thompson.
 *              Copyright (c) 2002 by Lucent Technologies.
 * Permission to use, copy, modify, and distribute this software for any
 * purpose without fee is hereby granted, provided that this entire notice
 * is included in all copies of any software which is or includes a copy
 * or modification of this software and in all copies of the supporting
 * documentation for such software.
 * THIS SOFTWARE IS BEING PROVIDED "AS IS", WITHOUT ANY EXPRESS OR IMPLIED
 * WARRANTY.  IN PARTICULAR, NEITHER THE AUTHORS NOR LUCENT TECHNOLOGIES MAKE ANY
 * REPRESENTATION OR WARRANTY OF ANY KIND CONCERNING THE MERCHANTABILITY
 * OF THIS SOFTWARE OR ITS FITNESS FOR ANY PARTICULAR PURPOSE.
 */

#include <stdarg.h>

#include "fmt/fmt.h"

int nothing(Fmt::State *s){
	return 0;
}

int nada(Fmt::State *s){
	return 0;
}

int
main(int argc, char *argv[])
{
	Fmt::quoteinstall();
	Fmt::print("hello world\n");
	Fmt::print("x: %x\n", 0x87654321);
	Fmt::print("u: %u\n", 0x87654321);
	Fmt::print("d: %d\n", 0x87654321);
	Fmt::print("s: %s\n", "hi there");
	Fmt::print("q: %q\n", "hi i'm here");
	Fmt::print("c: %c\n", '!');
	Fmt::print("g: %g %g %g\n", 3.14159, 3.14159e10, 3.14159e-10);
	Fmt::print("e: %e %e %e\n", 3.14159, 3.14159e10, 3.14159e-10);
	Fmt::print("f: %f %f %f\n", 3.14159, 3.14159e10, 3.14159e-10);
	Fmt::print("smiley: %C\n", (Rune)0x263a);
	Fmt::print("%g %.18g\n", 2e25, 2e25);
	Fmt::print("%2.18g\n", 1.0);
	Fmt::print("%f\n", 3.1415927/4);
	Fmt::print("%d\n", 23);
	Fmt::print("%i\n", 23);

	/* test %4$d formats */
	Fmt::print("%3$d %4$06d %2$d %1$d\n", 444, 333, 111, 222);
	Fmt::print("%3$d %4$06d %2$d %1$d\n", 444, 333, 111, 222);
	Fmt::print("%3$d %4$*5$06d %2$d %1$d\n", 444, 333, 111, 222, 20);
	Fmt::print("%3$hd %4$*5$06d %2$d %1$d\n", 444, 333, (short)111, 222, 20);
	Fmt::print("%3$lld %4$*5$06d %2$d %1$d\n", 444, 333, 111LL, 222, 20);

	/* test %'d formats */
	Fmt::print("%'d %'d %'d\n", 1, 2222, 33333333);
	Fmt::print("%'019d\n", 0);
	Fmt::print("%08d %08d %08d\n", 1, 2222, 33333333);
	Fmt::print("%'08d %'08d %'08d\n", 1, 2222, 33333333);
	Fmt::print("%'x %'X %'b\n", 0x11111111, 0xabcd1234, 12345);
	Fmt::print("%'lld %'lld %'lld\n", 1LL, 222222222LL, 3333333333333LL);
	Fmt::print("%019lld %019lld %019lld\n", 1LL, 222222222LL, 3333333333333LL);
	Fmt::print("%'019lld %'019lld %'019lld\n", 1LL, 222222222LL, 3333333333333LL);
	Fmt::print("%'020lld %'020lld %'020lld\n", 1LL, 222222222LL, 3333333333333LL);
	Fmt::print("%'llx %'llX %'llb\n", 0x111111111111LL, 0xabcd12345678LL, 112342345LL);

	/* test warning about multiple installation */
	Fmt::Formatter fmt;
	/* collide with pre-defined */
	fmt.print("should see warning about %%d\n");
	fmt.installverb('d', false, nothing);
	/* collide with custom */
	fmt.print("should see no warning about %%N\n");
	fmt.installverb('N', false, nothing);
	fmt.installverb('N', false, nothing); /* not a problem; redefining to same value. */
	fmt.print("should see one warning about %%N\n");
	fmt.installverb('N', false, nada);
	fmt.print("should see warning about %%alpha\n");
	fmt.installverb(0x3B1, false, nothing);
	fmt.installverb(0x3B1, false, nada);
	return 0;
}
