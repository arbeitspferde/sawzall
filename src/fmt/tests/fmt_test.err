Fmt: Warning: verb %N redefined
Fmt: Warning: verb %Î± redefined
