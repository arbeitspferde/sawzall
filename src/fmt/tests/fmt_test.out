hello world
x: 87654321
u: 2271560481
d: -2023406815
s: hi there
q: 'hi i''m here'
c: !
g: 3.14159 3.14159e+10 3.14159e-10
e: 3.141590e+00 3.141590e+10 3.141590e-10
f: 3.141590 31415900000.000000 0.000000
smiley: â˜º
2e+25 2e+25
 1
0.785398
23
23
111 000222 333 444
111 000222 333 444
111               000222 333 444
111               000222 333 444
111               000222 333 444
1 2,222 33,333,333
000,000,000,000,000
00000001 00002222 33333333
0,000,001 0,002,222 33,333,333
1111:1111 ABCD:1234 11:0000:0011:1001
1 222,222,222 3,333,333,333,333
0000000000000000001 0000000000222222222 0000003333333333333
000,000,000,000,001 000,000,222,222,222 003,333,333,333,333
0,000,000,000,000,001 0,000,000,222,222,222 0,003,333,333,333,333
1111:1111:1111 ABCD:1234:5678 110:1011:0010:0011:0101:0100:1001
should see warning about %d
should see no warning about %N
should see one warning about %N
should see warning about %alpha
